rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+
+* `edition`: A number that specifies the version of the flake
+  syntax/semantics to be used. This allows the interpretation of
+  flakes to change in the future. It also enables some evolution of
+  the Nix language; for example, the Nix files in the flake could be
+  parsed using a syntax determined by the edition. The only currently
+  allowed value is `201906`. Nix rejects flakes with an unsupported
+  edition.
+
+* `name`: An identifier for the flake, used to reference it from
+  `inputs`. (E.g. the `nixpkgs` in `inputs.nixpkgs` refers to the
+  identifier of the Nixpkgs flake.)
+
+* `description`: A short description of the flake.
+
+* `inputs`: The dependencies of the flake, as a list of flake
+  references (described below).
+
+* `outputs`: A function that, given an attribute set containing the
+  outputs of each of the input flakes keyed by their identifier,
+  yields the Nix values provided by this flake. Thus, in the example
+  above, `inputs.nixpkgs` contains the result of the call to the
+  `outputs` function of the `nixpkgs` flake, and
+  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
+  attribute of `nixpkgs`.
+
+  In addition to the outputs of each input, each input in `inputs`
+  also contains some metadata about the inputs. These are:
+
+  * `outPath`: The path in the Nix store of the flake's source
+    tree. This means that you could import Nixpkgs in a more
+    legacy-ish way by writing
+
+        with import inputs.nixpkgs { system = "x86_64-linux"; };
+
+    since `nixpkgs` still contains a `/default.nix`. In this case we
+    bypass its outputs entirely and only use the flake mechanism to
+    get its source tree.
+
+  * `rev`: The commit hash of the flake's Git repository.
+
+  * `revCount`: The number of ancestors of the revision `rev`. This is
+    not available for `github:...` repositories (see below), since
+    they're fetched as tarballs rather than as Git repositories.
+
+  * `lastModified`: The commit time of the revision `rev`, in the
+    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
+    this is available for both Git and GitHub repositories, so it's
+    useful for generating (hopefully) monotonically increasing version
+    strings.
+
+  The special input named `self` refers to the outputs and source tree
+  of *this* flake.
+
+  A number of outputs have a special meaning, as discussed below.
+
+## Well-known outputs
+
+A number of outputs have a specific meaning to Nix or other tools like
+Hydra. Currently, these are:
+
+* `packages`: A set of derivations used as a default by most `nix`
+  commands. For example, `nix run nixpkgs:hello` uses the
+  `packages.hello` attribute of the `nixpkgs` flake. It cannot contain
+  any non-derivation attributes. This also means it cannot be a nested
+  set! (The rationale is that supporting nested sets requires Nix to
+  evaluate each attribute in the set, just to discover which packages
+  are provided.)
```

Comment:
So things like `pythonPackages` won't work anymore?

Reply:
This is awkward for flakes that want to provide many outputs structured in some way, e.g. `haskellPackages`, `docs`, `developmentTools` etc. It sounds like currently those would have to be non-`packages` outputs, which reduces the usefulness of `packages` quite a lot.

Here's an alternative suggestion: allow users to provide an "index filter". Much like other filters, this would be applied at each step of looking for packages to decide whether to index the derivations inside that attribute set. The default would be "packages" (and not any child attribute sets!) but a user could specify a more liberal filter if they wanted (which would let you include attribute sets inside `packages`, or maybe even in other outputs).

Reply:
> supporting nested sets requires Nix to
> evaluate each attribute in the set, just to discover which packages
> are provided.

Why?

Reply:
When you encounter a attribute like `pythonPackages`, the only way to know that it's an attrset containing more packages rather than a package is by evaluating it.

Reply:
However turning `haskellPackages` into a flake itself would mean we don't need nesting right?

Reply:
There would likely be circular dependencies between `haskellPackages` and the rest of `nixpkgs` though.

Reply:
Given that flake UI already requires a complicated variable amount of colons, maybe have something like `subflakes` so that `github:NixOS/nixpkgs:pythonPackages:pygtk` could work?